### PR TITLE
Adding functionality `iterative_qpe`

### PR DESCRIPTION
**Pull request by KetpuntoG**
Adding functionality `iterative_qpe`

### CODE REVIEW

File: pennylane/ops/functions/__init__.py
Comment (Jaybsoni) [inline pennylane/ops/functions/__init__.py:48]: Add it to the doc-string above in this file

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:51]: The allowed types for estimation_wire don't include list? We should avoid it then in our examples
```suggestion
              measurements = qml.iterative_qpe(qml.RZ(2., wires = [0]), estimation_wire = 1, iters = 3)
```

File: tests/ops/functions/test_iterative_qpe.py
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:114]: Are probs and samples the only measurement processes currently supported? Maybe we should add a note in the function doc-string that tells users which final measurements are supported with IQPE?

File: tests/ops/functions/test_iterative_qpe.py
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:60]: Is there anyway to compare this with QPE to ensure the computed gradient makes sense?

File: tests/ops/functions/test_iterative_qpe.py
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:24]:
```suggestion
class TestIQPE:
    """Test to check that the iterative quantum phase estimation function works as expected."""
```

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:71]: Very clean 🥇!

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:22]: Maybe a question for @trbromley, should we add this to any of the major documentation pages? For example the functions subsection in [here](https://docs.pennylane.ai/en/stable/introduction/operations.html)

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:26]:
```suggestion
    Given a unitary matrix :math:`U`, this function applies the circuit for iterative quantum phase
    estimation and returns a list of the mid-circuit measurements with qubit reset.
```

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:38]: Maybe add a sentence explaining how to interpret the output. Also I think the PR which fixes mid-circuit measurement drawing has been merged? So we might be able to draw the circuit for this example as well.

**Review by Jaybsoni — COMMENTED** [review]
Amazing work! Looks pretty much good to go. I left some stylistic comments. The only other thing I can think to add if we have time would be one integration test which tests a full workflow like: given a hamiltonian compute its ground state energy using IQPE. 

Other than that, once the comments are addressed I am happy to approve 💯

File: pennylane/ops/functions/iterative_qpe.py
Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:38]: That PR does not consider `qml.cond` yet

File: tests/ops/functions/test_iterative_qpe.py
Comment (KetpuntoG) [reply tests/ops/functions/test_iterative_qpe.py:60]: I have not been able to get equal results. Being equivalent under certain conditions, I have the feeling that the gradients should not coincide. I think it is an interesting topic some researcher could study with PennyLane now :)

File: pennylane/ops/functions/__init__.py
Comment (KetpuntoG) [reply pennylane/ops/functions/__init__.py:48]: oh good catch

File: pennylane/ops/functions/iterative_qpe.py
Comment (mudit2812) [reply pennylane/ops/functions/iterative_qpe.py:38]: Yes, I'm in the process of adding support for `qml.cond`. Currently there's an open PR adding support for multiple `qml.cond` relying on the same (single) mid-circuit measurement, but that wouldn't be enough for drawing iterative QPE

File: tests/ops/functions/test_iterative_qpe.py
Comment (mudit2812) [inline tests/ops/functions/test_iterative_qpe.py:59]: There should also be tests for tf, torch, and autograd.

**Review by mudit2812 — COMMENTED** [review]
Thanks Guillermo! Such an exciting feature 😄 . Just one comment about testing gradients with other interfaces, but otherwise looks good. Once addressed, I'm ready to approve 🚀

File: pennylane/ops/functions/iterative_qpe.py
Comment (mudit2812) [inline pennylane/ops/functions/iterative_qpe.py:30]:
```suggestion
      estimation_wire (Union[Wires, int, str]): the wire to be used for the estimation.
```

**Comment by KetpuntoG** [comment]
I've added test for `tensorflow` and `torch`.
It has been identified a bug, so the tests will not pass until it is solved:
This is the issue:

https://github.com/PennyLaneAI/pennylane/issues/4825

File: pennylane/ops/functions/iterative_qpe.py
Comment (trbromley) [reply pennylane/ops/functions/iterative_qpe.py:22]: Yes it's a good point, I do feel like it's missing from somewhere. Yep we could put it [here](https://docs.pennylane.ai/en/stable/introduction/operations.html#operator-to-operator-functions).

File: tests/ops/functions/test_iterative_qpe.py
Comment (KetpuntoG) [reply tests/ops/functions/test_iterative_qpe.py]: updated with expval too 👍

File: pennylane/ops/functions/iterative_qpe.py
Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:38]: added example :)

File: pennylane/ops/functions/iterative_qpe.py
Comment (trbromley) [inline pennylane/ops/functions/iterative_qpe.py:25]:
```suggestion
    Given a unitary :math:`U`, this function applies the circuit for iterative quantum phase
```
Since we don't expect matrix input in this function.

**Comment by Jaybsoni** [comment]
[sc-45848]

File: doc/introduction/operations.rst
Comment (Jaybsoni) [inline doc/introduction/operations.rst:63]: Not blocking, just a suggestion for a future doc-string PR. This returns a measurement process (MP). Since we are starting to add more support for functions which return MP (qml.measure, qml.cond), we should add a new section. Something like `Operator to Measurement`

Comment (KetpuntoG) [reply doc/introduction/operations.rst:63]: It makes sense 👍
I would say that as soon as we have 2 of this type, we can make the section

Comment (KetpuntoG) [reply doc/introduction/operations.rst:63]: Or are you saying here that qml.measure and qml.cond is in the same family as iterative_qpe?

File: doc/releases/changelog-dev.md
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:38]: Now that I am seeing this, I wonder if this kind of return structure makes sense. Since we have 5 shots, I would think that it would be more intuitive if for each shot we had 3 measurement samples. This might apply to the whole of mid-circuit measurements. 

@mudit2812 @trbromley, what are your thoughts on this? Was this discussed in the PRD for MCMs?

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:48]: If this is not super important for the example, maybe we can remove for simplicity

File: pennylane/ops/functions/iterative_qpe.py
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:51]: Just a suggestion, we can rename the `estimation_wire` to be `"ancilla"` instead of 1; It's a little more clear for the user that its been added by the IQPE template and not just another wire in the circuit.

File: tests/ops/functions/test_iterative_qpe.py
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:124]: Are we planning to merge before this gets resolved?

**Review by Jaybsoni — APPROVED** [review]
Just some stylistic comments, otherwise I think this is good to go! I am approving for now, feel free to merge once the comments are resolved!

File: doc/introduction/operations.rst
Comment (Jaybsoni) [reply doc/introduction/operations.rst:63]: I am saying that `qml.measure` and `qml.cond` are in the same family, and we can make a new section now. But for now, I would say not to do it in this PR. But rather have a follow up PR to update docs. This way its not blocking the feature and we have more time to discuss with the product team.

File: doc/releases/changelog-dev.md
Comment (KetpuntoG) [reply doc/releases/changelog-dev.md:38]: Yes, this is what I mean when I talk about general statistics. It is something that will be included soon. At that time it will be possible to do it:

`qml.sample(measurements)` and it will return what you comment 👌

File: tests/ops/functions/test_iterative_qpe.py
Comment (KetpuntoG) [reply tests/ops/functions/test_iterative_qpe.py:124]: Yes, I think so, Matthew, who is working on the tensorflow fix, is aware of this test

File: pennylane/ops/functions/iterative_qpe.py
Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:51]: Okey, sounds good to me :)
I'll change it

File: pennylane/ops/functions/iterative_qpe.py
Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:48]: I prefer to leave it like this if you don't mind to emphasize that the initial state is not something that is included in the IQPE implementation

File: doc/releases/changelog-dev.md
Comment (KetpuntoG) [reply doc/releases/changelog-dev.md:25]: I prefer to leave it like this if you don't mind to emphasize that the initial state is not something that is included in the IQPE implementation

File: doc/releases/changelog-dev.md
Comment (KetpuntoG) [inline doc/releases/changelog-dev.md:28]:
```suggestion
    measurements = qml.iterative_qpe(qml.RZ(2., wires = [0]), ancilla = [1], iters = 3)
```

File: pennylane/ops/functions/iterative_qpe.py
Comment (soranjh) [inline pennylane/ops/functions/iterative_qpe.py:26]:
```suggestion
    estimation and returns a list of mid-circuit measurements with qubit reset.
```

File: pennylane/ops/functions/iterative_qpe.py
Comment (soranjh) [inline pennylane/ops/functions/iterative_qpe.py:30]:
```suggestion
      ancilla (Union[Wires, int, str]): the wire to be used for the estimation
```

File: pennylane/ops/functions/iterative_qpe.py
Comment (soranjh) [inline pennylane/ops/functions/iterative_qpe.py:60]:
```suggestion
    The output is an array of size ``(number of iterations, number of shots)``.
```

File: tests/ops/functions/test_iterative_qpe.py
Comment (mudit2812) [inline tests/ops/functions/test_iterative_qpe.py:123]: Since Matt's PR is merged now, does the test still need to xfail?

**Review by mudit2812 — COMMENTED** [review]
Looks good, just one question but otherwise I'm ready to approve :)

File: tests/ops/functions/test_iterative_qpe.py
Comment (KetpuntoG) [reply tests/ops/functions/test_iterative_qpe.py:123]: The error was discovered to be another. We had the possibility to make Torch or TF work and by preference Torch was chosen. @timmysilv  is aware of this

**Review by mudit2812 — APPROVED** [review]
Thanks!

File: pennylane/ops/functions/iterative_qpe.py
Comment (trbromley) [inline pennylane/ops/functions/iterative_qpe.py:67]: @KetpuntoG

> RZ(2.00)⁴⋅⁰

Is this `4.0` because the power is a float?

Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:67]: good catch! This was something we modified at the end to work with torch instead of tensorflow.
It is an integer. It is updated in this PR

https://github.com/PennyLaneAI/pennylane/pull/4871/files

**Pull request merged**
